### DISCUSSION TITLE
Can not insert foreign key into database? "message":"invalid input syntax for type bigint

### DISCUSSION THREAD

**Discussion by roker15** [supabase/supabase#3227] [Questions]
I have multiple tables related by foreign keys reffering to tables `ID`, each table ID is of type `int8`
When i tried to do
```js
const { data, error } = await supabase
  .from<Post>("posts")
  .insert([{ "post": content, subheading_id: { id: 1 } }]
  
``` 

I am getting following response 

```
{"hint":null,"message":"invalid input syntax for type bigint: \"{\"id\":1}\"","code":"22P02","details":null}
```

**Comment by MildTomato** [comment]
It looks like your 'subheading_id' column is a bigint (int8) type, but you're trying to update it like its a json column.
If you change it to what is below, you might find it will probably run.

```js
const { data, error } = await supabase.from<Post>("posts").insert([{ "post": content, subheading_id: 1 }])
```

It might help if you can explain in more detail what you're trying to do, and then maybe me or someone else here can advise

**Reply by roker15** [reply]
> It looks like your 'subheading_id' column is a bigint (int8) type, but you're trying to update it like its a json column.
> If you change it to what is below, you might find it will probably run.
> 
> ```js
> const { data, error } = await supabase.from<Post>("posts").insert([{ "post": content, subheading_id: 1 }])
> ```
> 
> It might help if you can explain in more detail what you're trying to do, and then maybe me or someone else here can advise

yes it does but then there is autocompletion error ---- because `subheading_id`  is a foreign key and the error is `Type 'number' is not assignable to type 'Subheading | undefined'`
Here is my Post object
`ype Post = {
    id: number,    
    created_at: string,
    updated_at: string,
    post: string,
    subheading_id:Subheading
};`

**Reply by MildTomato** [reply]
that's a typescript error, the `Subheading` type you have in `type Post` is probably not expecting a number

**Reply by MildTomato** [reply]
if you use the following, it will probably get things moving for you
`type Post = { id: number, created_at: string, updated_at: string, post: string, subheading_id: number };`

**Reply by roker15** [reply]
> that's a typescript error, the `Subheading` type you have in `type Post` is probably not expecting a number

That is not a typescript error, `subheading_id` is of type `Subheading` **_which is an object and not a number_**. so typescript is showing correct error. I think it is supabase error that it is not correctly able to reffer an object as foreign key. .... **I have done in MYSQL and JAVA --- you pass complete object of foreign key and it will save only the ID into database.**

**Reply by roker15** [reply]
> if you use the following, it will probably get things moving for you
> `type Post = { id: number, created_at: string, updated_at: string, post: string, subheading_id: number };`

But this is not the way. This will break the code and break the autocompletion everywhere when we fetch data. I think that this is bug.

**Reply by MildTomato** [reply]
> But this is not the way. This will break the code and break the autocompletion everywhere when we fetch data. I think that this is bug.

I'm not really sure about this, since if you fetched data from this table, subheading_id will always be a number.

However I can certainly see that this is confusing, me and @joshenlim have been looking through the docs to check and reproduce what you've been doing. It seems especially confusing with foreign keys and typescript types.

I think what we have to bear in mind is, when you select from a table, you could end up returning both the foreign key value, but also you could return values from the tables it has relations with. there's an example below of doing this.

It seems like we should be updating our docs to explain more [about using aliases in select queries](https://supabase.io/docs/reference/javascript/select#query-the-same-foreign-table-multiple-times), which seem really helpful when using typescript

What you could try is:

```js
type Subheading = {
  id: number,
  title: string,
  ..// anything else you like
}

type Post = {
  id: number,
  created_at: string,
  updated_at: string,
  post: string,
  subheading_id: number,
  subheading: Subheading
}

// insert example
const { data, error } = await supabase.from<Post>("posts").insert([{ "post": "content", subheading_id: 1 }]

// fetch example
const { data, error } = await supabase
  .from<Post>("posts")
  .select(`
    id,
    created_at,
    updated_at,
    post,
    subheading:subheading_id ( title ),
  `)
  .eq('id', 1) // filter to post that has id 1

```

**Comment by awalias** [comment]
hey @roker15 supabase-js works differently than a java ORM for mysql - inserting a nested object will not automatically be unpacked and stored in a foreign table

as the error (`"invalid input syntax for type bigint: \"{\"id\":1}\""`) suggests, supabase-js is expecting that you just pass the bigint value itself rather than an entire json object

so you'll probably want to insert the Subheading row initially, and then follow it up by inserting the Past row with a reference to the subheading

**Reply by roker15** [reply]
> hey @roker15 supabase-js works differently than a java ORM for mysql - inserting a nested object will not automatically be unpacked and stored in a foreign table
> 
> as the error (`"invalid input syntax for type bigint: \"{\"id\":1}\""`) suggests, supabase-js is expecting that you just pass the bigint value itself rather than an entire json object
> 
> so you'll probably want to insert the Subheading row initially, and then follow it up by inserting the Past row with a reference to the subheading

But then how we will model object on client side? if we model client side as per database , we have to assign foreign key as complete object and not as Int . If we model as `Int` how nested autocompletion will work? I think this is bad disign from supabase because i see this is working in every technology. 

You yourself is is providing automatic type generation in your documentation [here](https://supabase.io/docs/reference/javascript/generating-types#:~:text=Generate%20database%20types%20from%20OpenAPI%20specification%23) ..... so how will your types generated will work? Think about it... it's something bad design from firebase.

**Reply by MildTomato** [reply]
It sounds like using openapi-typescript you can get the type definitions of everything from your database and then, you could maybe construct typescript interfaces based on whatever you'd like to query.

It looks like openapi-typescript doesn't nest types for you based on foreign keys, but I guess you could do that yourself based on what you want to query.

_disclaimer - i've not tested any of this — please correct any of this if its wrong 😬 _

```js

// this is assuming you've been generating types and stored them somewhere
// also assumed that 'subheading' contains a column called 'title'
import { definitions } from "~/types/supabase";

// construct interface using definitions from openapi-typescript 
interface SelectQuery {
  extends <definitions['posts']> {
  subheading : definitions['subheading']
}

// fetch example
const { data, error } = await supabase
  .from<SelectQuery[]>("posts")
  .select(`
    *,
    subheading:subheading_id(title)
  `)
  .eq('id', 1) // filter to post that has id 1
  
/*  
would return something like 

data = {
  id: 1,
  created_at: "1631809187", 
  updated_at: 1631809187, 
  post: "some string", 
  subheading_id: 1
  subheader : {
    title: 'title of subheader'
  }
 }
 
 */
```

**Reply by roker15** [reply]
I have found an interesting workaround --- 
```
type posts={
....
subheading_id: Subheading | int
}
```
This way we can `post `as `int` and `get `as `Subheading `object so both will work without breaking autocompletion. you can also implement this pattern in your type generation tool. ....... Hope this will help other too

**Reply by roker15** [reply]
> It sounds like using openapi-typescript you can get the type definitions of everything from your database and then, you could maybe construct typescript interfaces based on whatever you'd like to query.
> 
> It looks like openapi-typescript doesn't nest types for you based on foreign keys, but I guess you could do that yourself based on what you want to query.
> 
> _disclaimer - i've not tested any of this — please correct any of this if its wrong 😬 _
> 
> ```js
> // this is assuming you've been generating types and stored them somewhere
> // also assumed that 'subheading' contains a column called 'title'
> import { definitions } from "~/types/supabase";
> 
> // construct interface using definitions from openapi-typescript 
> interface SelectQuery {
>   extends <definitions['posts']> {
>   subheading : definitions['subheading']
> }
> 
> // fetch example
> const { data, error } = await supabase
>   .from<SelectQuery[]>("posts")
>   .select(`
>     *,
>     subheading:subheading_id(title)
>   `)
>   .eq('id', 1) // filter to post that has id 1
>   
> /*  
> would return something like 
> 
> data = {
>   id: 1,
>   created_at: "1631809187", 
>   updated_at: 1631809187, 
>   post: "some string", 
>   subheading_id: 1
>   subheader : {
>     title: 'title of subheader'
>   }
>  }
>  
>  */
> ```

I have answered above my workaround -- please tell me if there is better way. 
my workaround
```
type posts={
....
subheading_id: Subheading | int
}
```
This way we can post as int and get as Subheading object so both will work without breaking autocompletion. you can also implement this pattern in your type generation tool.
